File: android/core/src/main/java/app/visly/shard/ShardRoot.kt
package app.visly.shard

import android.view.View
import androidx.annotation.Keep
import java.lang.RuntimeException

class ShardRoot(internal val ctx: ShardContext, @Keep internal val rustPtr: Long) {
    private fun finalize() { free() }
    private external fun free()
    private external fun getView(): ShardView
    private external fun measure(size: Size)

    fun measure(width: Float?, height: Float?): Size {
        val density = ctx.resources.displayMetrics.density
        measure(Size(width?.div(density) ?: Float.NaN, height?.div(density) ?: Float.NaN))

        fun updateFrame(root: ShardView) {
            val view = root.view
            view.layoutParams = AbsoluteLayout.LayoutParams(
                    root.frame.width().toInt(),
                    root.frame.height().toInt(),
                    root.frame.left.toInt(),
                    root.frame.top.toInt())

            if (view is AbsoluteLayout) {
                view.size = root.getSize()
                for (child in root.children) {
                    updateFrame(child)
                }
            }
        }

        updateFrame(getView())

        return getView().getSize()
    }

    internal val view: View by lazy {
        fun createHierarchy(root: ShardView): View {
            val view = root.view
            root.impl.bindView(view)

            if (view is AbsoluteLayout) {
                for (child in root.children) {
                    view.addView(createHierarchy(child))
                }
            } else if (root.children.size > 0) {
                throw RuntimeException("Only flexbox is allowed to specify children")
            }

            return view
        }

        createHierarchy(getView())
    }
}

File: ios/ShardKit/Classes/ShardRoot.swift
/**
 * Copyright (c) Visly Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import UIKit

public class ShardRoot {
    internal let context: ShardContext
    private let root: IOSRoot
    
    internal init(_ context: ShardContext, _ root: IOSRoot) {
        self.context = context
        self.root = root
    }
    
    deinit {
        shard_root_free(self.root)
    }
    
    public func measure(width: CGFloat?, height: CGFloat?) -> CGSize {
        shard_root_measure(root, CSize(width: Float(width ?? CGFloat.nan), height: Float(height ?? CGFloat.nan)), nil)
        let rootView: ShardView = Unmanaged.fromOpaque(UnsafeRawPointer(shard_root_get_view(root)!)).takeUnretainedValue()
        return rootView.size
    }
    
    func layout(width: CGFloat?, height: CGFloat?) -> CGSize {
        shard_root_measure(root, CSize(width: Float(width ?? CGFloat.nan), height: Float(height ?? CGFloat.nan)), nil)
        let rootView: ShardView = Unmanaged.fromOpaque(UnsafeRawPointer(shard_root_get_view(root)!)).takeUnretainedValue()
        
        func updateFrame(_ root: ShardView) {
            root.view.frame = root.frame
            root.impl.bindView(root.view)
            for child in root.children {
                updateFrame(child)
            }
        }

        updateFrame(rootView)
        return rootView.size
    }
    
    internal lazy var view: UIView = {
        let rootView: ShardView = Unmanaged.fromOpaque(UnsafeRawPointer(shard_root_get_view(root)!)).takeUnretainedValue()
        
        func createViewHierarchy(_ root: ShardView) -> UIView {
            let view = root.view
            if root.impl is FlexboxViewImpl {
                for child in root.children {
                    view.addSubview(createViewHierarchy(child))
                }
            } else if root.children.count > 0 {
                assertionFailure("Only flexbox is allowed to specify children")
            }
            
            return view
        }
        
        return createViewHierarchy(rootView)
    }()
}
